### PR TITLE
POM file version scanner updates and fixes

### LINKED ISSUE #39801: [BUG] pom_file_version_scanner.ps1 needs better errors around malformed lines
I mistakenly had a colon instead of a semicolon in a line in version_client.txt:
```
150:com.azure:azure-messaging-eventgrid;4.22.0;4.23.0-beta.1
151:com.azure:azure-messaging-eventgrid-namespaces:1.0.0-beta.1;1.0.0-beta.1
152:com.azure:azure-messaging-eventgrid-cloudnative-cloudevents;1.0.0-beta.1;1.0.0-beta.2
```

The resulting error confused me, because the line was obviously in the file:
```
➜ .\eng\versioning\pom_file_version_scanner.ps1
Using module powershell-yaml with version 0.4.7.
processing pomFile=D:\source\azure-sdk-for-java\sdk\eventgrid\azure-messaging-eventgrid-namespaces\pom.xml
Error: com.azure:azure-messaging-eventgrid-namespaces's dependency type is 'current' but the dependency does not exist in any of the version_*.txt files. Should this be an external_dependency? Please ensure the dependency type is correct or the dependency is added to the appropriate file.
There were errors encountered during execution. Please fix errors and run the script again.
This script can be run locally from the root of the repo. .\eng\versioning\pom_file_version_scanner.ps1
```

Ideally this would tell me I had a malformed line (missing a section) instead of telling me this.

### PR DESCRIPTION
**Pull request by JimSuplizio**
There are a few updates here:
1. #39801 - Additional checking when loading the version_* files, which are supposed to have 3 entries, separated by semicolons, consisting of groupId:artifactId**;**dependencyVersion**;**currentVersion. Right now, if someone creates a malformed entry, like having a colon instead of a semicolon in between the two versions, it doesn't error until it's scanning the POM files. Now, if the version_* files are being parsed, if there's no current version it'll check to see if the entry is an unreleased_ or beta_ entry and error if not.
2. If there's an error parsing any of the version_* files or the external dependencies, it'll error at that time instead of trying to scan the pom files with bad or incomplete data.
3. azure-cosmos-parent no longer exists, azure-cosmos-spark_3_2-12 should have been added as a valid parent so the other azure-comos-spark libaries were correctly scanned. CC: @kushagraThapar 
4. The $PomFilesIgnoreParent variable was valid 5 years ago when this was written. 4 years ago, the sdk/parents directory was restructured. This if statement, `if ($PomFilesIgnoreParent -contains $pomFile)`, was moot and didn't have any adverse effect. Note: This variable was actually used to ensure that azure-sdk-parent was getting scanned. This [one-line change in the script ensures](https://github.com/Azure/azure-sdk-for-java/pull/43243/files#diff-8cd53d5081f3ba5f2b79e30602b41248da54c48f8492a1f5eb6ab780020e1bd3R594) that it is now getting scanned.
5. The last thing here was adding verification for [plugin->executions->configuration->signature](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/parents/azure-client-sdk-parent/pom.xml#L318) which was something brand new. @alzimmermsft and @billwert. This is the only item I have any concern over, I am unfamiliar with this and it's unclear if someone can add a signature that doesn't use a library.

Fixes #39801 

I'm running [java - template - ci](https://dev.azure.com/azure-sdk/public/_build/results?buildId=4389382&view=results) and [java - cosmos - ci](https://dev.azure.com/azure-sdk/public/_build/results?buildId=4389381&view=results) against the refs/merge of this PR. Note: Only the Analyze job runs the Verify versions in POM files step which runs this script.

### CODE REVIEW

File: eng/versioning/pom_file_version_scanner.ps1 [hunk 1]
```diff
@@ -558,25 +589,18 @@ Get-ChildItem -Path $Path -Filter pom*.xml -Recurse -File | ForEach-Object {
         return
     }
 
-    if ($PomFilesIgnoreParent -contains $pomFile)
+    $xmlPomFile = New-Object xml
+    $xmlPomFile.Load($pomFile)
+    if ($ValidParents -notcontains $xmlPomFile.project.parent.artifactId)
```
Comment (billwert) [inline eng/versioning/pom_file_version_scanner.ps1:594]: I think we should probably scan the parent POMs also. I did this in the branch I was working on:
```
+    if ($ValidParents -notcontains $xmlPomFile.project.parent.artifactId -and $ValidParents -notcontains $xmlPomFile.project.artifactId)
```

Comment (JimSuplizio) [reply eng/versioning/pom_file_version_scanner.ps1:594]: @billwert, funny is that this is the same change I'd made locally and will be up with the next commit. Just an FYI the only parent this isn't currently scanning is [sdk/parents/azure-sdk-parent/pom.xml](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/parents/azure-sdk-parent/pom.xml) which is effectively the grandparent and, itself, doesn't have a parent. All of the other ValidParents, aside from azure-sdk-parent,
`$ValidParents = ("azure-sdk-parent", "azure-client-sdk-parent", "azure-data-sdk-parent", "azure-perf-test-parent", "azure-cosmos-spark_3_2-12")`
were getting scanned. In the case of azure-client-sdk-parent and azure-data-sdk-parent, their parent is azure-sdk-parent, and for azure-perf-test-parent and azure-cosmos-spark_3_2-12 their parent is azure-client-sdk-parent.

File: eng/versioning/pom_file_version_scanner.ps1 [hunk 2]
```diff
@@ -794,6 +818,64 @@ Get-ChildItem -Path $Path -Filter pom*.xml -Recurse -File | ForEach-Object {
...
+            $isPartOfConfig = Confirm-Node-Is-Part-Of-Configuration $signatureNode
+            if (!$isPartOfConfig)
+            {
+                $hasError = $true
+                # Because this particular case is harder to track down, print the OuterXML which is effectively the entire tag
+                $potentialLogMessage = Join-With-NewLine $potentialLogMessage "Error: signature is missing version element and/or artifactId and groupId elements signatureNode=$($signatureNode.OuterXml)"
+            }
+            continue
+        }
+        # plugins should always have an artifact but may not have a groupId
```
Comment (billwert) [inline eng/versioning/pom_file_version_scanner.ps1:838]:
```suggestion
        # signatures should always have an artifact but may not have a groupId
```

File: eng/versioning/pom_file_version_scanner.ps1 [hunk 3]
```diff
@@ -794,6 +818,64 @@ Get-ChildItem -Path $Path -Filter pom*.xml -Recurse -File | ForEach-Object {
...
+                {
+                    $hasError = $true
+                    $potentialLogMessage = Join-With-NewLine $potentialLogMessage "$($retVal)"
+                }
+            }
+        }
+        else
+        {
+            $hasError = $true
+            $potentialLogMessage = Join-With-NewLine $potentialLogMessage "Error: Missing plugin version update tag for groupId=$($groupId), artifactId=$($artifactId). The tag should be <!-- {x-version-update;$($groupId):$($artifactId);current|dependency|external_dependency<select one>} -->"
```
Comment (billwert) [inline eng/versioning/pom_file_version_scanner.ps1:875]:
```suggestion
            $potentialLogMessage = Join-With-NewLine $potentialLogMessage "Error: Missing signature version update tag for groupId=$($groupId), artifactId=$($artifactId). The tag should be <!-- {x-version-update;$($groupId):$($artifactId);current|dependency|external_dependency<select one>} -->"
```

File: eng/versioning/pom_file_version_scanner.ps1 [hunk 4]
```diff
@@ -510,13 +519,35 @@ $ArtifactsPerSD = [ordered]@{};
...
+# exit. There's no point in scanning the pom files if any of the files have invalid or duplicate entries.
 $libHash = @{}
-Build-Dependency-Hash-From-File $libHash $Path\eng\versioning\version_client.txt $false
-Build-Dependency-Hash-From-File $libHash $Path\eng\versioning\version_data.txt $false
+$verClientFile = Join-Path $Path "eng/versioning/version_client.txt"
+Build-Dependency-Hash-From-File $libHash $verClientFile $false
+if ($script:FoundError)
+{
+    Write-Error-With-Color "There were errors encountered building the dependency hash from version_client.txt. Please fix errors and run the script again."
+    Write-Error-With-Color "This script can be run locally from the root of the repo. .\eng\versioning\pom_file_version_scanner.ps1"
```
Comment (billwert) [inline eng/versioning/pom_file_version_scanner.ps1:530]: maybe jam this in a function

**Review by billwert — APPROVED** [review]
left a couple suggestions but lgtm!

File: eng/versioning/pom_file_version_scanner.ps1 (hunk 4 above)
Comment (JimSuplizio) [reply eng/versioning/pom_file_version_scanner.ps1:530]: At this point, I'm just going to leave this as is.

**Comment by JimSuplizio** [comment]
> left a couple suggestions but lgtm!

I committed all of the suggestions with updates.

**Comment by JimSuplizio** [comment]
Nothing automatically runs against changes to this file. As per the description I've run a couple of pipelines to test and the results are there. I'm going to merge.

**Pull request merged**